### PR TITLE
Configuring Jenkins (Master, Agents, Pipelines), Maintain map zoom after sidebar is closed

### PR DESCRIPTION
**Pull request by goemen**

### CODE REVIEW

File: openshift/3.11/pipelines/Jenkinsfile.cicd [hunk 1]
```diff
@@ -29,7 +29,7 @@ pipeline {
     API_DIRECTORY = "backend"
     FRONTEND_DIRECTORY = "frontend"
     MAINTENANCE_DIRECTORY = "maintenance"
-    DEVOPS_DIRECTORY = "openshift/3.11"
+    DEVOPS_DIRECTORY = "openshift/4.0"
```
Comment (Fosol) [inline openshift/3.11/pipelines/Jenkinsfile.cicd:32]: You shouldn't need to update this as it is in the `/3.11/` folder.  There should be a new file in the `/4.0/` folder with the update.  We'll clean out 3.11 once everything works.

File: openshift/4.0/pipelines/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- Deploys the resulting artifacts to the `DEV` environment.
+- _(TODO) Triggers an asynchronous OWASP Security Scan on the deployed application._
+
+### Test (`promote-to-test-pipeline`)
+
+> :bulb: This section is in progress
+
+This manually triggered pipeline promotes deployments from the `dev` environment to the `test` environment. To perform the promotion click the **Start Pipeline** on the `promote-to-test-pipeline`.
+
+### Prod (`promote-to-prod-pipeline`)
```
Comment (Fosol) [inline openshift/4.0/pipelines/README.md:23]: Missing the `cicd-pipeline` that builds the `master` branch and deploys it to the **TEST** environment.  This should be essentially identical to the `dev` to **DEV**.

Comment (Fosol) [reply openshift/4.0/pipelines/README.md:23]: Also remember when this pipeline from `master` goes to **TEST** you need to ask for the user to input the version number to tag the image.

File: openshift/4.0/scripts/oc-sonar-dotnet.sh [hunk 3]
```diff
@@ -30,17 +30,17 @@ export PATH="$PATH:/root/.dotnet/tools"
...
+BACKEND_DIR="${BACKEND_DIR:-../../../backend}"
 TEST_DIR="${TEST_DIR:-${BACKEND_DIR}/tests/unit}"
 COVERAGE_DIR="${COVERAGE_DIR:-${TEST_DIR}/TestResults}"
 
 # Check requirements
 #
-dotnet tool list -g | grep sonarscanner >/dev/null 2>&1 || {
-  fatal_error "sonar scanner for msbuild not installed on not in PATH"
-}
+# dotnet tool list -g | grep sonarscanner >/dev/null 2>&1 || {
```
Comment (Fosol) [inline openshift/4.0/scripts/oc-sonar-dotnet.sh:41]: Devin will need to get this part working in his PR

File: openshift/4.0/templates/jenkins-slaves/dotnet-jenkins-slave.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      <volumes/>
+      <containers>
+        <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
+          <name>jnlp</name>
+          <image>registry.access.redhat.com/dotnet/dotnet-31-jenkins-slave-rhel7:latest</image>
+          <privileged>false</privileged>
+          <alwaysPullImage>true</alwaysPullImage>
+          <workingDir>/tmp</workingDir>
+          <command></command>
+          <args>${computer.jnlpmac} ${computer.name}</args>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/dotnet-jenkins-slave.yaml:27]: Odd to see parameters but no defined parameters in this template

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-config-map.yaml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+objects:
+  - apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: "${NAME}"
+      labels:
+        role: jenkins-slave
+    data:
+      podTemplate: |-
+        <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-config-map.yaml:20]: Odd to have Alejandro's namespace here?

File: openshift/4.0/templates/jenkins-slaves/dotnet-jenkins-slave.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      <inheritFrom></inheritFrom>
+      <name>dotnet-31</name>
+      <instanceCap>5</instanceCap>
+      <idleMinutes>0</idleMinutes>
+      <label>dotnet-31</label>
+      <serviceAccount>jenkins</serviceAccount>
+      <nodeSelector></nodeSelector>
+      <volumes/>
+      <containers>
+        <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/dotnet-jenkins-slave.yaml:20]: Odd to see Alejandro's namespace here?

File: frontend/src/components/maps/leaflet/Map.tsx [hunk 7]
```diff
@@ -215,8 +234,19 @@ const Map: React.FC<MapProps> = ({
   }, [interactive, mapRef]);
 
   const handleMapFilterChange = async (filter: IPropertyFilter) => {
-    setGeoFilter(getQueryParams(filter));
-    setChanged(true);
+    const compareValues = (objValue: any, othValue: any) => {
+      return whitelistedFilterKeys.reduce((acc, key) => {
+        return (
+          (_.isEqual(objValue[key], othValue[key]) ||
```
Comment (devinleighsmith) [inline frontend/src/components/maps/leaflet/Map.tsx:240]: this is a nitpick, but you import isEqual specifically from lodash but end up just using _ in the end.

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-generic-template.yaml [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        - type: ImageChange
+  - apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: "${NAME}"
+      labels:
+        role: jenkins-slave
+    data:
+      node10: |-
+        <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-generic-template.yaml:60]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-generic-template.yaml [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+          <name>${NAME}</name>
+          <label>${NAME}</label>
+          <instanceCap>2</instanceCap>
+          <idleMinutes>5</idleMinutes>
+          <activeDeadlineSeconds>600</activeDeadlineSeconds>
+          <serviceAccount>jenkins</serviceAccount>
+          <nodeSelector></nodeSelector>
+          <volumes/>
+          <containers>
+            <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-generic-template.yaml:71]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-npm/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    -p NAME=jenkins-slave-npm \
+    -p SOURCE_CONTEXT_DIR=openshift/4.0/templates/jenkins-slaves/jenkins-slave-npm \
+    | oc apply -f -
+```
+
+For all params see the list in the `../jenkins-slave-generic-template.yaml` or run `oc process --parameters -f ../jenkins-slave-generic-template.yaml`.
+
+## Jenkins
+
+Add a new Kubernetes Container template called `jenkins-slave-npm` (if you've build and pushed the container image locally) and specify this as the node when running builds. If you're using the template attached; the `role: jenkins-slave` is attached and Jenkins should automatically discover the slave for you. Further instructions can be found [here](https://docs.openshift.com/container-platform/3.11/using_images/other_images/jenkins.html#using-the-jenkins-kubernetes-plug-in).
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-npm/README.md:28]: Change `3.11` to `4.0`

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-dotnet/README.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    -p BUILDER_IMAGE_NAME=registry.redhat.io/dotnet/dotnet-31-jenkins-agent-rhel7 \
+    -p SOURCE_CONTEXT_DIR=openshift/4.0/templates/jenkins-slaves/jenkins-slave-dotnet \
+    | oc apply -f -
+```
+
+For all params see the list in the `../jenkins-slave-generic-template.yaml` or run `oc process --parameters -f ../jenkins-slave-generic-template.yaml`.
+
+## Jenkins
+
+Add a new Kubernetes Container template called `jenkins-slave-dotnet` (if you've build and pushed the container image locally) and specify this as the node when running builds. If you're using the template attached; the `role: jenkins-slave` is attached and Jenkins should automatically discover the slave for you. Further instructions can be found [here](https://docs.openshift.com/container-platform/3.11/using_images/other_images/jenkins.html#using-the-jenkins-kubernetes-plug-in).
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-dotnet/README.md:29]: Change `3.11` to `4.0`

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-zap/Jenkinsfile.test [hunk 12]
```diff
@@ -0,0 +1,17 @@
+pipeline {
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/jenkins-slave-zap/Jenkinsfile.test:1]: What's this file for?

File: openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml [hunk 13]
```diff
@@ -0,0 +1,39 @@
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: nodejs-10-jenkins-slave
+  labels:
+    role: jenkins-slave
+  namespace: 354028-tools
+data:
+  node10: |-
+    <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml:10]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      <inheritFrom></inheritFrom>
+      <name>nodejs-10</name>
+      <instanceCap>5</instanceCap>
+      <idleMinutes>0</idleMinutes>
+      <label>nodejs-10</label>
+      <serviceAccount>jenkins</serviceAccount>
+      <nodeSelector></nodeSelector>
+      <volumes/>
+      <containers>
+        <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml:20]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml [hunk 15]
```diff
@@ -0,0 +1,44 @@
+apiVersion: v1
+data:
+  podTemplate: |-
+    <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml:4]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+      <name>owasp-zap</name>
+      <label>owasp-zap</label>
+      <instanceCap>2</instanceCap>
+      <idleMinutes>5</idleMinutes>
+      <activeDeadlineSeconds>600</activeDeadlineSeconds>
+      <serviceAccount>jenkins</serviceAccount>
+      <nodeSelector></nodeSelector>
+      <volumes/>
+      <containers>
+        <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
```
Comment (Fosol) [inline openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml:15]: Odd to have Alejandro's namespace here

File: openshift/4.0/templates/maintenance/caddy.bc.yaml [hunk 17]
```diff
@@ -1,6 +1,6 @@
 ---
 kind: Template
-apiVersion: v1
+apiVersion: template.openshift.io/v1
```
Comment (Fosol) [inline openshift/4.0/templates/maintenance/caddy.bc.yaml:3]: Was this required?

**Review by Fosol — COMMENTED** [review]
Missing README on how to setup the four pipelines

- `dev` branch to **DEV** (automated)
- **DEV** to **TEST** (manual - just deploy existing image)
- `master` branch to **TEST** (automated - requires version tag to entered)
- **TEST** to **PROD** (manual - deploys specified version tagged image, also does DB backup)

**Review by devinleighsmith — APPROVED** [review]

File: openshift/4.0/scripts/oc-sonar-dotnet.sh (hunk 3 above)
Comment (goemen) [reply openshift/4.0/scripts/oc-sonar-dotnet.sh:41]: Yes, requires sonarqube deployment

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-config-map.yaml (hunk 5 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/jenkins-slave-config-map.yaml:20]: I dont think this is Alejandro.

File: openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml (hunk 13 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml:10]: https://github.com/jenkinsci/kubernetes-plugin

File: openshift/4.0/templates/maintenance/caddy.bc.yaml (hunk 17 above)
Comment (goemen) [reply openshift/4.0/templates/maintenance/caddy.bc.yaml:3]: Not required

File: openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml (hunk 16 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml:15]: Not Alejandro

File: openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml (hunk 15 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/zap-jenkins-slave.yaml:4]: Not Alejandro

File: openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml (hunk 14 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/node10-jenkins-slave.yaml:20]: Not Alejandro

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-npm/README.md (hunk 10 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/jenkins-slave-npm/README.md:28]: 4.1

File: openshift/4.0/templates/jenkins-slaves/jenkins-slave-zap/Jenkinsfile.test (hunk 12 above)
Comment (goemen) [reply openshift/4.0/templates/jenkins-slaves/jenkins-slave-zap/Jenkinsfile.test:1]: Copied the file from the 3.11

File: openshift/4.0/pipelines/README.md (hunk 2 above)
Comment (goemen) [reply openshift/4.0/pipelines/README.md:23]: Will need to verify this with you @Fosol tomorrow

File: openshift/4.0/pipelines/Jenkinsfile.cicd [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+    // To enable pipeline verbose debug output set to "true"
+    DEBUG_OUTPUT = sh(script: 'echo "${DEBUG_OUTPUT:-false}"', returnStdout: true).trim()
+  }
+  stages {
+    stage('Initialize') {
+      steps {
+        script {
+          // load supporting functions from external script files
+          commonPipeline = load "${DEVOPS_DIRECTORY}/pipelines/scripts/common.groovy"
+          notify = load "${DEVOPS_DIRECTORY}/pipelines/scripts/notify.groovy"
```
Comment (agalbrai) [inline openshift/4.0/pipelines/Jenkinsfile.cicd:66]: all the "groovy"s makes me think of Austin Powers haha

File: openshift/4.0/pipelines/scripts/notify.groovy [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+// ------------------
+// Notifications
+// ------------------
+
+def success(appName, version, environment, rc_url = "") {
+  rocketChatSend(
+    rc_url,
+    "${appName}, release: [${version}] was deployed to **${environment}** 🚀",
+    "#1ee321",  // green
+    "https://i.imgur.com/zHzPWXr.png"
```
Comment (agalbrai) [inline openshift/4.0/pipelines/scripts/notify.groovy:15]: opened this link to see what it was and chuckled, I forgot about this

**Review by Fosol — APPROVED** [review]

**Pull request merged**